components/browser/session/src/main/java/mozilla/components/browser/session/SessionObserver.kt

```diff
@@ -0,0 +1,58 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.session
+
+/**
+ * This class is a combination of [Session.Observer] and
+ * [SessionManager.Observer]. It provides functionality to observe changes to a
+ * specified or selected session, and automatically takes care of switching
+ * over the observer in case a different session is selected (see [observe] and
+ * [observeSelected]).
+ *
+ * @property activeSession the currently observed session
+ * @property sessionManager the application's session manager
+ */
+abstract class SessionObserver(
+    private val sessionManager: SessionManager
+) : SessionManager.Observer, Session.Observer {
+
+    protected open var activeSession: Session? = null
+
+    /**
+     * Starts observing changes to the specified session.
+     *
+     * @param session the session to observe.
+     */
+    fun observe(session: Session) {
+        activeSession = session
+        session.register(this)
+    }
+
+    /**
+     * Starts observing changes to the selected session (see
+     * [SessionManager.selectedSession]). If a different session is selected
+     * the observer will automatically be switched over and only notified of
+     * changes to the newly selected session.
+     */
+    fun observeSelected() {
+        activeSession = sessionManager.selectedSession
+        sessionManager.register(this)
+        activeSession?.register(this)
+    }
+
+    /**
+     * Stops the observer.
+     */
+    fun stop() {
+        sessionManager.unregister(this)
+        activeSession?.unregister(this)
+    }
+
+    override fun onSessionSelected(session: Session) {
+        activeSession?.unregister(this)
+        activeSession = session
+        session.register(this)
+    }
```

Comment:
I'm currently reviewing the feature-downloads patch and I wonder if there's a race condition here: Can `stop()` and `onSessionSelected()` be called in such an order that we are still registered to a session?

Reply:
@pocmo So, In the current design, I don't see a race condition:

- Opening the TabsTray stops the BrowserFragment which stops the ToolbarFeature->Presenter ->SessionObserver
-  A new session is selected in the tabs tray -> the ToolbarPresenter no longer observes that
- The TabsTrayFragment stops and the BrowserFragment starts again re-registering the observer

So, no race condition, but it also means that `onSessionSelected` might not even be needed? Should we not start/stop the ToolbarPresenter in this case?

Reply:
Marking this as resolved as per our chat. Even if an app doesn't use fragments and therefore uses `onSessionSelected` there still won't be a race conditions as all methods in `ObserverRegistry` are synchronized on `observers`.